### PR TITLE
No Weight Decay Keywords

### PR DESCRIPTION
**Pull request by francesco-bertolotti**
Regarding the issue [https://github.com/pytorch/torchtitan/issues/2599](https://github.com/pytorch/torchtitan/issues/2599):

I’ve added a parameter group that excludes weight decay, which is populated based on the `no_weight_decay_keywords` configuration option. Currently, `no_weight_decay_keywords` is a comma-separated list of glob patterns.

For example, setting it to `*emb*,*norm*,*out*` will include in the no-weight-decay parameter group the token embeddings, all normalization layer weights, and the output layer weights.

### CODE REVIEW

File: torchtitan/components/optimizer.py [hunk 1]
```diff
@@ -78,6 +80,10 @@ class Config(Configurable.Config):
         weight_decay: float = 0.1
         """Weight decay to use"""
 
+        no_weight_decay_keywords: str = ""
```
Comment (tianyu-l) [inline torchtitan/components/optimizer.py]: Do you think it makes sense to make this general?
E.g. we can make this a `dict[str, int]` from fqn re to their `weight_decay_multiplier`. The number of keys would correspond to the number of parameter groups.

File: torchtitan/components/optimizer.py [hunk 2]
```diff
@@ -4,9 +4,10 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
+import fnmatch
```
Comment (tianyu-l) [inline torchtitan/components/optimizer.py]: should we use `re` which is more powerful and more commonly used in torchtitan?

File: torchtitan/components/optimizer.py
Comment (tianyu-l) [inline torchtitan/components/optimizer.py:281]: @fegin if we have multiple parameter group, what should we do here?

File: torchtitan/components/optimizer.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/components/optimizer.py]: should the default be not applying any special multiplier, e.g. to embedding / norm / lm_head?

Comment (francesco-bertolotti) [reply torchtitan/components/optimizer.py]: I’ve mostly seen weight decay applied selectively (i.e., excluded for certain parameters) rather than using different scaling multipliers, but your idea could be an interesting way to generalize this behavior.

That said, we should design it in a way that integrates cleanly with the CLI. One option could be to allow repeated flags, for example:

```bash
--weight_decay_scale *norm* 0
--weight_decay_scale *emb* 0
--weight_decay_scale *out* 0
```

What do you think?

File: torchtitan/components/optimizer.py (hunk 2 above)
Comment (francesco-bertolotti) [reply torchtitan/components/optimizer.py]: I understand your concern. However, if our goal is to select parameters using bash-style globbing, I think `fnmatch` is the better fit, it already provides the matching behavior we need without relying on custom regular expressions, which can be more error-prone.

If we decide to support a different syntax instead of bash-style globbing, then using another approach would make sense. Also, both `re` and `fnmatch` are part of the standard library, so adopting `fnmatch` wouldn’t introduce any additional dependencies.

That said, if using `fnmatch` isn’t an option, I’m happy to switch to `re` without any issues.

File: torchtitan/components/optimizer.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/components/optimizer.py]: Yeah I noticed that idea might break CLI, too. Let me think about it.

Comment (tianyu-l) [reply torchtitan/components/optimizer.py]: @francesco-bertolotti I discussed with @fegin and here are the latest thoughts

> I’ve mostly seen weight decay applied selectively (i.e., excluded for certain parameters) rather than using different scaling multipliers

We've definitely seen people using more than no-weight-decay, i.e. different parameters having different scaling multipliers, so that's real use case.

> That said, we should design it in a way that integrates cleanly with the CLI.

It is somewhat similar to parameter init which is being refactored in https://github.com/pytorch/torchtitan/pull/2633. In general it feels to me that CLI is not the right place to config them.

If users really want to use CLI to toggle on/off weight decay, they can introduce a translate layer (e.g. by subclassing `OptimizersContainer`) to support CLI on top of torchtitan.

WDYT?

File: torchtitan/components/optimizer.py
Comment (fegin) [reply torchtitan/components/optimizer.py:281]: This should work if `all_params` is the parameter group list like what is done in this PR.

File: torchtitan/components/optimizer.py [hunk 3]
```diff
@@ -111,16 +117,45 @@ def _build_optimizer_kwargs(config: Config) -> dict[str, Any]:
...
+                        for pattern in config.no_weight_decay_keywords.split(",")
+                    )
+                ]
+                logger.info(
+                    f"Not applying weight decay to parameters {[n for n,p in nowd_params]} in model {model._get_name()}"
+                )
+                optimizer.add_param_group(
+                    {"params": [p for n, p in nowd_params], **optimizer_nowd_kwargs}
+                )
+                all_params.extend(p for n, p in nowd_params)
```
Comment (fegin) [inline torchtitan/components/optimizer.py]: We should make this a generic parameter group logic instead of hard-coding the weight decay here. People may use parameter groups for optimization choices. So as @tianyu-l mentions above, it seems to us that the generic way is to specify the parameter optimization options in the config similar to parameter init then you should be able to use a loop to categorize the parameters into different groups.

File: torchtitan/components/optimizer.py (hunk 1 above)
Comment (francesco-bertolotti) [reply torchtitan/components/optimizer.py]: I have looked at https://github.com/pytorch/torchtitan/pull/2633 and it seems like a more general way to approach this. I will try to adopt the same solution.

**Comment by francesco-bertolotti** [comment]
Hi @tianyu-l and @fegin,

This PR refactors the optimizer configuration system to support per-parameter-group overrides, enabling fine-grained control over optimizer hyperparameters (e.g., assigning different learning rates or weight decay to specific parameter groups such as embeddings).

While working on this, the scope expanded slightly beyond hyperparameter overrides: the new system also supports using different optimizer types within the same run. This allows assigning distinct optimizers to different parameter groups, which can be particularly useful for cases like using `SparseAdam` for embedding layers.

---

### **Key Changes**

#### **1. New `BaseOptimizer` abstraction** (`torchtitan/components/optimizer.py`)

* Introduces a `BaseOptimizer` class with a nested `Config` dataclass.
* Adds concrete subclasses (`AdamW`, `Adam`) that replace the previous string-based optimizer resolution (`_resolve_optimizer_cls`).
* Each optimizer subclass:

  * Defines its own config schema (`lr`, `betas`, `eps`, `weight_decay`, `fused`, `foreach`)
  * Implements `get_optimizer_kwargs()` to construct optimizer arguments
* Supports config composition via `BaseOptimizer.__or__`, enabling clean layering of overrides on top of defaults.

---

#### **2. `OptimizersContainer` config overhaul**

* Replaces the previous flat config (`name`, `lr`, `beta1`, `beta2`, etc.) with a structured schema:

```python
default: BaseOptimizer.Config
overrides: dict[str, BaseOptimizer.Config]
```

* `default` defines the base optimizer configuration applied to all parameters.

* `overrides` enables per-parameter configuration using parameter name matching.

* Parameters are grouped based on override keys, and separate optimizer parameter groups are constructed with the appropriate configs.

* `OptimizersInBackwardContainer` has also been updated to follow this new structure.

---

#### **3. Config registry migrations (16 files)**

* Updated all model and experiment configs (e.g., `llama3`, `llama4`, `deepseek_v3`, `flux`, `gpt_oss`, `qwen3`, as well as `ft`, `rl`, `vlm`) from the flat syntax:

```python
OptimizersContainer.Config(lr=8e-4)
```

to the new nested format:

```python
OptimizersContainer.Config(
    default=AdamW.Config(lr=8e-4)
)
```

---

#### **4. Tests**

* Updated unit tests (`tests/unit_tests/test_train_spec.py`) to:

  * Use the new `Adam.Config` structure
  * Reflect the updated access pattern (`optimizers[0][0]`)

---

### **Example Usage**

```python
OptimizersContainer.Config(
    default=AdamW.Config(lr=3e-4, weight_decay=0.1),
    overrides={
        "model.tok_embeddings.weight": AdamW.Config(
            lr=1e-4,
            weight_decay=0.0,
        ),
    },
)
```

---

### **Results**

Loss curves before and after these changes:

vsopts

**Comment by tianyu-l** [comment]
I'm giving a different try in https://github.com/pytorch/torchtitan/pull/2859. Could you help review?
